Added beta specs docs (#1977)
holochain/holochain: docs/specs/src/hwp_1_intro.md

Reviewer: instead of:
Hashes: All data on our DHT
maybe:
Hashes: All data on a Holochain DHT

so that we aren't accidentally indicating to people that Holochain has one DHT.

Reviewer: instead of:
Hashes: All data on our DHT...
maybe:
Hashes: All data on a Holochain DHT...

so that we aren't accidentally indicating to people that Holochain has one DHT.

Reviewer: typo:
MARSHALL PEASE, ROBERT SHOSTAK, and LESLIE LAMPORT, this single data reality is called "interactive consistency" as is about the vector of "Private Values" sent by each node.
    
    change to "and is about the vector ..."

took on a very formal approach to describing generalized distributed
computation schemes and carefully compared the approach taken by
Blockchain based systems with the approach taken by Holochain as well as
providing formal reasoning for the benefits of Holochain's agent centric
approach.

In this second version, we describe Holochain on its own terms by
providing the highest level view of our approach and our starting
axioms. At the next level of detail we provide an informal system
description using the metaphor of games. Then we dive into deeper detail
with a formal system description including the specific context and
assumptions we come from, the integrity guarantees Holochain offers, and
a formal state model and discussion of security and safety concerns.
Finally at the most detailed level we provide an implementation
specification in Appendix A.

Axioms
------

We begin by stating our axioms[^1] regarding the nature of coordination:

1.  **Coordination arises from agents** starting from the same
  ground-rules and acting as soon as each party in any interaction
  can confirm that each other\'s actions confirm that actions or

>   interactions conform to those ground rules. (Thus, in our frame,

Reviewer 2: Having reflected about this for a while I'm thinking that in "real, fuzzy life" there are gradients in how subjectively held rules map onto each other. And coordination starts (or fails) to arise, as soon as at least one agent subjectively concludes based on past behavior of the other agents involved that the gradients between his subjectively held rules and those of the other agents involved are small enough that its worth laying an initial "trace of actions" for others to pick up on.

That Holochain, on the layer of the software code, embodies a means for *each* party in *any* interaction to be able to confirm that each other's actions confirm that actions or interactions conform to those ground rules is then rather a specific attempt for a solution to this more fuzzy problem of how to get agents start playing together according to shared rules. But I think I don't consider it to be appropriate to take as an axiom.
For example, it may start to show its limitations when "non-techie" people interact with Holochain that have no internal maps whatsoever that they could use to subjectively verify for themselves with high level of confidence how and whether those rules are *actually* embodied in the software.

>   interactions conform to those ground rules. (Thus, in our frame,

Reviewer 3: > to confirm that each other's actions confirm that actions or interactions conform to those ground rules

Hmm, there's that clause again, so the fact that you're phrasing it this way makes me think I'm missing something and not able to parse it. Diving into the English grammar of this clause: what does it mean for "actions to confirm that actions conform"? The similarity of "confirm" and "conform" notwithstanding, this seems confusingly self-referential, and maybe that's on purpose, but it's breaking my brain.

>   interactions conform to those ground rules. (Thus, in our frame,

Author: I see that there's a typo there.  I think it should read:

```
**Coordination arises from agents** starting from the same
  ground-rules and any single agent's ability to act as soon as they have 
  confirmed to their satisfaction that other agents existing actions
  conform to those ground rules.
```

>   interactions conform to those ground rules. (Thus, in our frame,

Author: clairfied

  coordination looks like alignment heading in the same direction,
  rather than agreement proceeding in lockstep.)

> 2.  **Coordination is grammatic**, in that it comes from embodying a

Reviewer 4: I understand you want to keep using it. I think it deserves an introduction or explanation. Even though I can make sense of the word by itself, I don't understand it in this context.

> 2.  **Coordination is grammatic**, in that it comes from embodying a

Author: footnote added to do this explanation

>   geometry that removes uncertainty, and embodying an ability to

Author: added a footnote

>   geometry that removes uncertainty, and embodying an ability to

Reviewer 3: @zippy which footnote? I would expect to see a superscript number somewhere around here, linking to the relevant footnote.

  compose different coordinative subsystems that have different

>   ground-rules.

Reviewer 4: Why different ground rules? The subsystems share the ground rules, as stated further down for social coherence (authority), that all agents know the ground rules.

Author: Each DNA is a set of ground-rules.  The overall system embodies being able to compose them.

Axiom 1 arises from the insight that we cannot fight against the
physical reality of different experiences by different nodes in the
context of networked interaction nor fight against the real-world
impossibility of determining global temporal order because of the
non-existence of simultaneity. Instead, any coordination system must
align its ontology with the truth that **global state actually does not
exist**. Thus we start with what does actually exist: local temporal
state. This local state can be shared with, and validated by, others in
conformity with pre-defined ground-rules. In so doing we can still
achieve difficult and complex coordination safely (including in the
context of problems on the scale of global monetary transactions)
without the costs and bottlenecks that arise from starting from the
ontology of a single shared global state. Holochain is an implementation
of a system using this alternate frame.

Axiom 2 arises from the insight that systems for successful large scale
coordination demand the property of anti-fragility, that is, they must

> perform better under perturbation[^2]. Coordination happens in the

Author: I think in anti-fragility the idea is that it's not just about consistency of performance, it's improvement.  Where a fragile thing breaks under perturbation, and anti-fragile thing gets better.

> perform better under perturbation[^2]. Coordination happens in the

Reviewer 4: Got it

context of fundamentally dynamic environments in which the coordinating
elements are changed by the fact of their coordination. Coordination is
a co-evolutionary context. We claim by this axiom that what meets the
challenge of anti-fragility in such contexts is composable sub-systems,

> in which the composition comes out of a grammatics that embodies the

Reviewer 3: ```suggestion
in which the composition comes out of a grammar that embodies the
```

Reviewer 3: I see you resolved this, which I assume means you prefer "grammatics", but I wonder what it means, since it's not a word I see used anywhere else. Is it like "application of grammars"?

Author: Yes, I'm trying to introduce this notion of "grammatic capacites"  I'm using that specifically because "grammar" by itself makes people think of language, and "grammatic" is my way of generalizing about that.  I will add a footnote.

Author: ```
FOOTNOTE for "grammatic"
We use the term grammatic as any a way to indicate the general pattern of what grammar and individual grammars are an instance of.  Where grammar is often understood to be limited to language, grammatics points to the pattern of creating templates with classes of items that can fill those templates.  This pattern can be used for creating "grammars" of social interaction,  "grammars" of physical structures (we would call Alexander's "Pattern Language" for architecture a particular instance of grammatics) and so on. 
```

actual dimensionality of the problem subdomains (i.e. their geometry),
and by which agents in that context can react powerfully to
perturbations because the available composability is dimensionally

> aligned.

Reviewer 3: Here's that bit on embodying geometry again, with a bit more explanation. I still feel I have to guess at what this means. It is clearly a complex and nuanced thought and the writing could use more explanation, or a reference to some other material.

Is this just saying that the ground rules can be written in an unambiguous way using language (grammar) which fully specifies the problem (sub)domain without room for interpretation? Is it something like, "the map IS the territory"?

Author: Well, I'm not actually talking about that determinism.  I'm talking about anti-fragility being met along a gradient by living inside grammatic contexts that actually align with the dimensionality of the space the grammar(s) are about.  English (and all human languages) align grammatically with the context of meaning making.  But it's a gradient.  For example, they fail quite miserably by introducing a grammatic element (the verb to-be) which actually doesn't align with dimensionality of reality of meaning in that that projects on to the outside world something that's actually interior, i.e. "false isness" and hence E-Prime.  So in this regard the gradient of alignment  of the grammar to "reality" is off.  But this is too much to talk about in this intro.

Reviewer 3: This is profound, especially since you're asserting it's at the heart of Holochain design, but I wonder how effective it will be to say this without more explanation. I am left really wanting to know more about what this means, but with no way to find out. Without breaking this down or pointing to some other resource, it may come across as some tantalizing but half-baked thought which it's clearly not. Maybe you could write an appendix on this? I feel like this deserves at least the promise of more detail later, if not the actual deliverance of more detail here and now.

Reviewer 4: Big +1
As Michael said we know there's some real genius behind the Holochain design. But that phrase with "grammatics", "dimensionality", "problem subdomains", "their geometry" and "dimensionally aligned" is a riddle wrapped up in an enigma. It may also be mistaken as a string of lofty words by someone who doesn't want to take the time to understand it.

Author: To me a whitepaper is exactly a place where one defines "jargon".  So I guess I need to do that for these terms.  I will add more footnotes.  See above for one on grammatics...

Author: I added some footnotes, this may make it worse!

Reviewer 3: Are the footnotes meant to explain what "grammatic embodiment" and "dimensional alignment" mean? The nearby footnotes reference The Players of Ludos and Antifragile, which don't seem to explain these terms.

I really think there needs to be some extra context to support this jargon. I'm not against jargon. There just needs to be a resource for whoever wants to dive in to understand it (including myself), rather than skimming over it.

The core axiom (though not explicitly stated as such) of the Byzantine
Generals Problem is that coordination starts AFTER "consensus on state",
i.e. that the Lieutenants can't execute their plan until they have
followed the consensus algorithm and arrived at single data reality[^3].
This axiom leads system designers to figure out how to implement
machinery for **Global Consensus**. Our axioms lead us, instead, to
implement tooling for **Scaling Consent** as an alternate solution to

> the Byzantine Generals Problem.

Reviewer 3: This is solid and profound!

From Global Consensus to Scaled Consent
---------------------------------------

In distributed systems, we believe that it is absolutely fundamental to
understand that every action taken by an agent in any social context,
happens because that agent has crossed a confidence threshold of some
sort, from its own point of view, that the given action is appropriate

> to take. Stated another way: agentic assessment of the social context

Reviewer 3: ```suggestion
to take. Stated another way: agent-centric assessment of the social context
```

Is "agentic" synonymous with "agent-centric"? I think if you use it, it would help to introduce it as such. I see a few references of each throughout the documents.

Reviewer 3: Ah, I didn't realize "agentic" is a real word:

> That behaves like an agent: able to express or expressing agency or control on one's own behalf or on the behalf of another

Reviewer 4: Where did you find that, @maackle?
I just found
```
agentive | ˈājən(t)iv |
adjective
formal taking an active role; producing an effect: the program envisions designers as agentive, social citizens.
• Grammar denoting a word or part of a word that expresses the doer of an action: agentive nouns are typically derived from a verb through suffixation of -er.
```

Reviewer 3: Well, on Wiktionary. But also I see it used in some psychological contexts.

Author: I like agentic more than agentive.  I think more people will resonate with it.  But I could be wrong.

to provide confidence for action, which is always contextual. Such
confidence indicates **social coherence**. This notion of social
coherence is the single most important design goal of Holochain: to
create the tooling that in contextually appropriate ways leads to
increasing social coherence.

Given the above, we propose a very simple approach to creating tooling
capacity for building increasing certainty: **enable validated global
visibility, on demand, of local state**. In this approach, we
distinguish between authorship, which is about local state changes
initiated by agents, and authority, which is about distributing the
responsibility of validating and making visible those state changes
across the participants in the network. This approach requires that we:

1.  Ensure that all agents know the "ground-rules". i.e. offer a
  framework for composing lots of small units of social agreement in
  which players can add deterministic elements of certainty into
  their interactions.

2.  Ensure that all agents can *reliably* see what's going on. i.e.
  offer a framework for adding to and querying a collectively held
  database in which there is a knowable level of confidence in its

>   reliability and integrity that can also be graduated to fit the

Reviewer 4: "that can be graduated to fit the circumstances' - what does that mean?

Author: It means in the sense of the definition: "change (something, typically color or shade) gradually or step by step."  That's a less common meaning of the word, maybe something else would be better?

Reviewer 4: I understand "graduated", but the "knowable level of confidence in its reliability and integrity" reads as very defined whereas "graduated to fit the circumstances" by itself without further details waters down the strictness IMO

Reviewer 3: Maybe if it adds something like "graduated to fit the circumstances, ranging from an arbitrarily low level of confidence to the highest possible level certainty"

Author: Added what you suggested @maackle

  circumstances.

The first point we deliver through various types of **Validation
Rules**. Validation rules create certainty in the following dimensions
with examples:

-   Content: a value does not exceed a maximum length

-   Structure: an entry consists of a certain set of types of data

-   Sequence: someone can not spend credits they have not already
  received earlier on their chain

-   Process: a transaction must be approved and signed by a notary

-   Behavior: one does not take an action more frequently than a certain
  rate

The second point we deliver through various types of **Intrinsic Data
Integrity**. We use a number of cryptographic methods to create
self-proving data of various types:

> -   Provenance: An agent's network address is their public-key. Thus,

Reviewer 4: ```suggestion
-   Provenance: An agent's network address is their public key. Thus,
```

  when interacting with agents it's possible to have deterministic
  confidence in who one is interacting with because there is no
  identity layer between network locations subject to attack

>   surface. I.e. Unlike the web address, you don't need a domain name

Reviewer 4: ```suggestion
  surface. I.e. unlike the web address, you don't need a domain name
```

  certificate associated with the domain name to become confident of
  "who" you are talking to.

-   Signatures: Because provenance is a public key, it's also easy to
  create self-proving authenticity. All messages sent, and all data
  committed to chains is signed by agents using their public key.
  Thus any agent can immediately and with high confidence, verify
  the authenticity of messages and data.

-   Hashes: All data on our DHT is addressed by its hash. Thus, when
  retrieving data it's possible to have deterministic confidence
  that it hasn't been tampered with by whoever was storing it.

> -   Structural Monotonicity (for chains and DHT): non-repudiable

Reviewer 4: repudiable exists in Spanish, what is this in English? non-discardable?

Author: un-deniable

  state-change \[TODO: ACB\]

-   Logical Monotonicity: Crossings that you can't uncross (once
  something has become true, it can't become untrue) \[TODO: ACB\]

-   Common Genesis: The Validation Rules and joining criteria of an

>   application are the first first entry in every chain. This

Reviewer 4: ```suggestion
  application are the first entry in every chain. This
```

  provides a mechanism for self-proving shared ground rules. Any

>   agent can examine the chain of any other agent all the way back to

Reviewer 4: Any agent can inspect any other agent's source chain? Isn't that private?

Author: Nope the chain can be inspected on the DHT quite easily, just query backwards down the action-hashes. You might not get the content of private entries but the chain is always available.
